src/Microsoft.ML.Core/Data/VBufferMutationContext.cs

```diff
@@ -0,0 +1,148 @@
+            return destination.GetMutableContext(
+                newLogicalLength,
+                valuesCount,
+                keepOldOnResize: keepOldOnResize,
+                requireIndicesOnDense: requireIndicesOnDense);
+        }
+
+        internal static VBufferMutationContext<T> Create<T>(
+            ref VBuffer<T> destination,
+            int newLogicalLength,
+            int valuesCount,
+            int maxValuesCapacity)
+        {
+            return destination.GetMutableContext(
+                newLogicalLength,
+                valuesCount,
+                maxValuesCapacity);
+        }
+    }
+
+    /// <summary>
+    /// An object capable of mutation a <see cref="VBuffer{T}"/> by filling out
+    /// <see cref="Values"/> (and <see cref="Indices"/> if the buffer is not dense).
+    /// </summary>
+    public ref struct VBufferMutationContext<T>
+    {
+        private readonly int _logicalLength;
+        private readonly T[] _values;
+        private readonly int[] _indices;
+
+        /// <summary>
+        /// The mutable span of values.
+        /// </summary>
+        public readonly Span<T> Values;
+
+        /// <summary>
+        /// The mutable span of indices.
+        /// </summary>
+        public readonly Span<int> Indices;
+
+        /// <summary>
+        /// Gets a value indicating whether a new Values array was allocated.
+        /// </summary>
+        public bool CreatedNewValues { get; }
+
+        /// <summary>
+        /// Gets a value indicating whether a new Indices array was allocated.
+        /// </summary>
+        public bool CreatedNewIndices { get; }
+
+        internal VBufferMutationContext(int logicalLength,
+            int physicalValuesCount,
+            T[] values,
+            int[] indices,
+            bool requireIndicesOnDense,
+            bool createdNewValues,
+            bool createdNewIndices)
+        {
+            _logicalLength = logicalLength;
+            _values = values;
+            _indices = indices;
+
+            bool isDense = logicalLength == physicalValuesCount;
+
+            Values = _values.AsSpan(0, physicalValuesCount);
+            Indices = !isDense || requireIndicesOnDense ? _indices.AsSpan(0, physicalValuesCount) : default;
+
+            CreatedNewValues = createdNewValues;
+            CreatedNewIndices = createdNewIndices;
+        }
+
+        /// <summary>
+        /// Creates a new <see cref="VBuffer{T}"/> using the current
+        /// Values and Indices.
+        /// </summary>
+        /// <param name="physicalValuesCount">
+        /// An optional size that allows reducing the number of physical values to be
+        /// represented in the created buffer.
+        /// This is useful in sparse situations where the mutation context was created
+        /// with a larger physical value count than was needed
+        /// because the final value count was not known at creation time.
+        /// </param>
+        /// <returns></returns>
```

Comment:
><returns></returns> [](start = 12, length = 19)

remove empty

Reply:
fixing.

---
In reply to: [233651783](https://github.com/dotnet/machinelearning/pull/1580#discussion_r233651783) [](ancestors = 233651783)